### ISSUE TITLE
Add training optimizations to train.py

### ISSUE THREAD

**Issue by pvonderlind** [pvonderlind/TinyLM#3]
Speed up training by introducing performance optimizations like:
* Half-Precision training to BFloat16
* Model compilation to achieve Kernel Fusion for all non Self-Attention layers.
* Adding FlashAttention to achieve Kernel Fusion in the Self-Attention layers.
* Gradient Clipping
* Add Fused AdamW if available to achieve Kernel Fusion in the optimizer.
